plugins/kinetic-dp/fu-kinetic-dp-puma-aux-isp.c

```diff
@@ -0,0 +1,863 @@
+	(guint32)(value[3] | ((guint32)value[2] << 8) | ((guint32)value[1] << 16) |                \
+		  ((guint32)value[0] << 24))
+
+typedef enum {
+	PUMA_CHIP_RESET_REQUEST = 0,
+	PUMA_CODE_LOAD_REQUEST = 0x01,
+	PUMA_CODE_LOAD_READY = 0x03,
+	PUMA_CODE_BOOTUP_DONE = 0x07,
+	PUMA_CMDB_GETINFO_REQ = 0xA0,
+	PUMA_CMDB_GETINFO_READ = 0xA1,
+	PUMA_CMDB_GETINFO_INVALID = 0xA2,
+	PUMA_CMDB_GETINFO_DONE = 0xA3,
+	PUMA_FLASH_ERASE_DONE = 0xE0,
+	PUMA_FLASH_ERASE_FAIL = 0xE1,
+	PUMA_FLASH_ERASE_REQUEST = 0xEE,
+	PUMA_FW_UPDATE_DONE = 0xF8,
+	PUMA_FW_UPDATE_READY = 0xFC,
+	PUMA_FW_UPDATE_REQUEST = 0xFE,
+} AuxWinModeRequestType;
+
+typedef enum {
+	PUMA_MODE_CHUNK_PROCESSED = 0x03,
+	PUMA_MODE_CHUNK_RECEIVED = 0x07,
+	PUMA_MODE_FLASH_INFO_READY = 0xA1,
+	PUMA_MODE_UPDATE_ABORT = 0x55,
+} AuxWinModeStatusType;
+
+typedef struct {
+	guint8 ObjectType;
+	guint8 ObjectSubType;
+	guint8 ObjectLength[sizeof(guint32)];
+} HeaderInfoFormat;
+
+struct _FuKineticDpPumaAuxIsp {
+	FuKineticDpAuxIsp parent_instance;
+};
+
+typedef struct {
+	guint32 isp_processed_size;
+	guint32 isp_total_size;
+	guint16 read_flash_prog_time;
+	guint16 flash_id;
+	guint16 flash_size;
+} FuKineticDpPumaAuxIspPrivate;
+
+G_DEFINE_TYPE_WITH_PRIVATE(FuKineticDpPumaAuxIsp,
+			   fu_kinetic_dp_puma_aux_isp,
+			   FU_TYPE_KINETIC_DP_AUX_ISP)
+
+#define GET_PRIVATE(o) (fu_kinetic_dp_puma_aux_isp_get_instance_private(o))
+
+static gboolean
+fu_kinetic_dp_puma_aux_isp_enter_code_loading_mode(FuKineticDpConnection *connection,
+						   GError **error)
+{
+	guint8 cmd;
+	guint8 status;
+	/*this may need adjust or as input parameter instead*/
+	guint32 max_time_ms = PUMA_CODE_LOAD_READY_MAX_WAIT;
+	/*this may need adjust or as input parameter instead*/
+	guint32 interval_ms = POLL_INTERVAL_MS;
+
+	g_debug("Entering Puma isp driver code loading mode...");
+	/* send PUMA_CODE_LOAD_REQUEST cmd */
+	cmd = PUMA_CODE_LOAD_REQUEST;
+	if (!fu_kinetic_dp_connection_write(connection,
+					    PUMA_DPCD_SINK_MODE_REG,
+					    &cmd,
+					    sizeof(cmd),
+					    error)) {
+		g_prefix_error(
+		    error,
+		    "failed to write PUMA_DPCD_SINK_MODE_REG with PUMA_CODE_LOAD_REQUEST: ");
+		return FALSE;
+	}
+
+	/* Wait for the command to be processed */
+	while (max_time_ms != 0) {
+		if (!fu_kinetic_dp_connection_read(connection,
+						   PUMA_DPCD_SINK_MODE_REG,
+						   &status,
+						   1,
+						   error)) {
+			g_prefix_error(error,
+				       "failed to read PUMA_DPCD_SINK_MODE_REG for status: ");
+			return FALSE;
+		}
+		if (status == PUMA_CODE_LOAD_READY) {
+			return TRUE;
+		} else {
+			/* wait interval before check the status again */
+			g_usleep((gulong)(interval_ms * 1000));
+			if (max_time_ms >= interval_ms)
+				max_time_ms -= interval_ms;
+			else
+				max_time_ms = 0;
+		}
+	}
+	/* time out */
+	g_set_error_literal(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INTERNAL,
+			    "Waiting PUMA_CODE_LOAD_READY timed-out");
```

Comment:
lowercase `Waiting` please -- we chain these together so sentence case doesn't really make sense

Reply:
fixed